cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+	}
+	sort.Strings(writeSlice)
+	fileio.WriteToFileHandle(writer, "construct\tcounts")
+	for _, i := range writeSlice {
+		fileio.WriteToFileHandle(writer, i)
+	}
+}
+
+// parseBam takes in a cellranger count bam file and pulls out reads that are representative of the UMI and also returns the construct associated with the UMI.
+func parseBam(s Settings) {
+	var k int = 0
+	var constructName, cellString, cellByConstructName, umiBx string
+	var count float64
+	var found bool
+	var bit int32
+	var norm bool = false
+	var sc bool = false
+	var noSettings bool = false
+	var allConstructs, cellTypeSlice, umiBxSlice []string
+
+	if s.inputNormalize != "" { //create a bool for normalize
+		norm = true
+	}
+	if s.scAnalysis != "" { //create a bool for cellTypeAnalysis
+		sc = true
+	}
+	if !sc && s.binCells < 1 && !s.samOut && !s.byCell && !s.umiSat {
+		noSettings = true
+	}
+
+	ch, head := sam.GoReadToChan(s.inFile)
+
+	out := fileio.EasyCreate(s.outFile)
+
+	if s.samOut {
+		sam.WriteHeaderToFileHandle(out, head)
+	}
+
+	var tree = make([]interval.Interval, 0)
+	var selectTree map[string]*interval.IntervalNode
+	if s.bed != "" {
+		bedEntries := bed.Read(s.bed)
+		for _, i := range bedEntries {
+			tree = append(tree, i)
+		}
+		selectTree = interval.BuildTree(tree)
+	}
+
+	pseudobulkMap := make(map[string]float64)
+
+	for i := range ch { //iterate of over the chanel of sam.Sam
+		if s.umiSat {
+			umi, _, _ := sam.QueryTag(i, "UB")
+			cb, _, _ := sam.QueryTag(i, "CB")
+			umiBx = fmt.Sprintf("%s_%s", umi, cb)
+			umiBxSlice = append(umiBxSlice, umiBx)
+		}
+		num, _, _ := sam.QueryTag(i, "xf") //xf: extra flags (cellranger flags)
+		bit = num.(int32)
+		if bit&8 == 8 { // bit 8 is the flag for a UMI that was used in final count. I call these "valid" UMIs.
+			k++
+			if s.bed != "" {
+				overlappedBedEntry := interval.Query(selectTree, i, "any")
+				if len(overlappedBedEntry) != 1 {
+					continue
+				}
+				constructName = overlappedBedEntry[0].(bed.Bed).Name
+			} else {
+				construct, _, _ := sam.QueryTag(i, "GX") // get the construct associated with valid UMI
+				constructName = construct.(string)
+			}
+			if s.byCell || sc || s.binCells > 0 || s.determineBins != "" {
+				cell, _, _ := sam.QueryTag(i, "CB") // get the cell barcode associated with valid UMI
+				cellString = cell.(string)
+				cellByConstructName = fmt.Sprintf("%s\t%s", cellString, constructName) //list of all construct reads and what cell barcodes they belong to
+				if s.byCell {                                                          //if you only want to print out the construct and the cell it was found in
+					fileio.WriteToFileHandle(out, cellByConstructName)
+				}
+				if sc || s.binCells > 0 || s.determineBins != "" { //get some data for the singleCellAnalysis and distributeCells functions
+					allConstructs = append(allConstructs, constructName) //list of all constructs found in the bam
+					cellTypeSlice = append(cellTypeSlice, cellByConstructName)
```

Comment:
I thought cellTypeSlice is a []string of cell types. Is cellByConstructName a string of cell type?

Reply:
cellByConstructName is maybe poorly named that is the slice that has cellBx \t construct